File: C++/0368-largest-divisible-subset.cpp
// Time:  O(n^2)
// Space: O(n)

class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        if (nums.empty())
            return {}; // or return vector<int>();
        vector<int> dp (nums.size(), 1);
        vector<int> prev(nums.size(), -1);
        sort(nums.begin(), nums.end());
        int largest_idx = 0;
        for (int i = 0; i < nums.size(); ++i) {
            for (int j = 0; j < i; ++j) {
                if (nums[i] % nums[j] == 0 && dp[j] + 1 > dp[i]) {
                    dp[i] = dp[j] + 1;
                    prev[i] = j;
                }
            }
            if (dp[largest_idx] < dp[i]) {
                largest_idx = i;
            }
        }
        vector<int> result;
        for (int i = largest_idx; i != -1; i = prev[i]) {
            result.push_back(nums[i]);
            //or .emplace
        }
        reverse(result.begin(), result.end());
        return result;
    }
};


// Notes:
// dp[i]: the size of the largest distinct subset of the first
// i+1 numbers including nums[i]

File: Python/0368-largest-divisible-subset.py
# Time:  O(n^2)
# Space: O(n)
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        if not nums:
            return []
        nums.sort()
        dp = [1] * len(nums)
        prev = [-1] * len(nums)
        largest_idx = 0
        for i in range(len(nums)):
            for j in range(i):
                if nums[i] % nums[j] == 0:
                    if dp[i] < dp[j] + 1:
                        dp[i] = dp[j] + 1
                        prev[i] = j
            if dp[largest_idx] < dp[i]:
                largest_idx = i

        result = []
        i = largest_idx
        while i != -1:
            result.append(nums[i])
            i = prev[i]
        return result[::-1]


# Notes:
# -Variant of Largest Increasing Subset
# -Cond: Current element is divisible by previous elements (since sorted)
# -If not sorted, check both a%b and b%a
# -if b%a = 0 & c%b = 0 then c%a = 0 (Transitive property)
# -dp[i] denotes the length of largest subset including the current element
# nums[i]
